gym_ignition/utils/inverse_kinematics_nlp.py

```diff
@@ -0,0 +1,490 @@
+# Copyright (C) 2020 Istituto Italiano di Tecnologia (IIT). All rights reserved.
+# This software may be modified and distributed under the terms of the
+# GNU Lesser General Public License v2.1 or any later version.
+
+import os
+import numpy as np
+from enum import Enum, auto
+from typing import List, Dict, Optional, Union, NamedTuple
+
+
+class TargetType(Enum):
+    POSITION = auto()
+    ROTATION = auto()
+    TRANSFORM = auto()
+
+
+class TransformTargetData(NamedTuple):
+    position: np.ndarray
+    quaternion: np.ndarray
+
+
+class TargetData(NamedTuple):
+    type: TargetType
+    weight: float
+    data: Union[np.ndarray, TransformTargetData]
+
+
+class IKSolution(NamedTuple):
+    joint_configuration: np.ndarray
+    base_position: np.ndarray = np.array([0.0, 0.0, 0.0])
+    base_quaternion: np.ndarray = np.array([1.0, 0.0, 0.0, 0.0])
+
+
+class RotationParametrization(Enum):
+    QUATERNION = auto()
+    ROLL_PITCH_YAW = auto()
+
+    def to_idyntree(self):
+        import iDynTree
+
+        if self.value == RotationParametrization.QUATERNION.value:
+            return iDynTree.InverseKinematicsRotationParametrizationQuaternion
+        elif self.value == RotationParametrization.ROLL_PITCH_YAW.value:
+            return iDynTree.InverseKinematicsRotationParametrizationRollPitchYaw
+        else:
+            raise ValueError(self.value)
+
+
+class TargetResolutionMode(Enum):
+    TARGET_AS_CONSTRAINT_FULL = auto()
+    TARGET_AS_CONSTRAINT_NONE = auto()
+    TARGET_AS_CONSTRAINT_POSITION = auto()
+    TARGET_AS_CONSTRAINT_ROTATION = auto()
+
+    def to_idyntree(self):
+        import iDynTree
+
+        if self.value == TargetResolutionMode.TARGET_AS_CONSTRAINT_FULL.value:
+            return iDynTree.InverseKinematicsTreatTargetAsConstraintFull
+        elif self.value == TargetResolutionMode.TARGET_AS_CONSTRAINT_NONE.value:
+            return iDynTree.InverseKinematicsTreatTargetAsConstraintNone
+        elif self.value == TargetResolutionMode.TARGET_AS_CONSTRAINT_POSITION.value:
+            return iDynTree.InverseKinematicsTreatTargetAsConstraintPositionOnly
+        elif self.value == TargetResolutionMode.TARGET_AS_CONSTRAINT_ROTATION.value:
+            return iDynTree.InverseKinematicsTreatTargetAsConstraintRotationOnly
+        else:
+            raise ValueError(self.value)
+
+
+class InverseKinematicsNLP:
+
+    def __init__(self, urdf_filename: str, controlled_joints: List[str] = None) -> None:
```

Comment:
As far as I understand from the signature, this IK method does not support kinematic chains with (redundant) non-fixed non-actuated joints. Is this a limitation of iDyntree's NLP IK or of the interface?

Reply:
Sorry for the missing documentation. It could be very useful particularly for this class but I think I'll delay it to another PR due to time constraints.

The default behaviour is performing IK on all joints of the model. This mode is enabled if you do not pass the `controlled_joints` argument (honestly, this argument should be called `considered_joints` or `enabled_joints`). Instead, if you pass a list of joints, the solver only optimizes over them, converting the remaining to fixed joints.

Reply:
> I'll delay it to another PR due to time constraints

Ok

> The default behaviour is performing IK on all joints of the model. This mode is enabled if you do not pass the controlled_joints argument (honestly, this argument should be called considered_joints or enabled_joints). Instead, if you pass a list of joints, the solver only optimizes over them, converting the remaining to fixed joints.

I see, now it's more clear, thank you.
I also think that `considered_joints`, or `enabled_joints`, would be better namings.

Reply:
I renamed the option to `considered_joints`